### DISCUSSION TITLE
Problem with edge classification on graph

### DISCUSSION THREAD

**Discussion by JHZhou123** [pyg-team/pytorch_geometric#2433] [Q&A]
Hi Matthias,

I am working on a problem that the dataset is different graphs and the label is one of the existing edge  in each graph, so I try to use GNN to train the graphs and use 

```
out=out[row]+out[col]
F.log_softmax(out, dim=1)
```

to find the edge with the highest probability, but in this case global pooling is not used, so how to use mini-batching ?

I also find out if the model predict the label (one edge of the graph) directly, the result is good. But if I train the graph, upgrade node attribute and use the mlp(node1+node2) to find out the edge  with the highest  probability, the model is not working and the result is poor, maybe this method does not make sense. Can you help me?

thank you!

**Comment by rusty1s** [comment]
It's tricky to implement that in a mini-batch fashion, as you need to ensure that your positive link exists in the mini-batch. I guess you could perform subgraph sampling, where you ensure that your positive link exists in the subgraph, and then proceed with your pipeline as usual.

**Reply by JHZhou123** [reply]
Do you have any idea that why the model is not learning well when I try to output one edge with the highest probability from the whole edge set?

**Reply by rusty1s** [reply]
I'm not really sure what that means TBH. Your training method should look similar to:
```python
out = out[row] + out[col]
prob = F.softmax(out, dim=0)
loss = F.cross_entropy(prob[edge_idx], torch.tensor([1])
```

**Reply by JHZhou123** [reply]
I use
`out = out[row] + out[col]`
out is [edge_number,1]
then
```
log_prob = F.log_softmax(out, dim=0)
loss=F.nll_loss(out.T,edge_index)
```
but the performance is poor, is this a good way to do it?

**Reply by rusty1s** [reply]
No, that does not make much sense from my point of view. If you want to find the most important edge from a set of edges, then you should do something like I mentioned above (where `edge_idx` represents the index of the ground-truth edge).

**Reply by JHZhou123** [reply]
Thank you, but do you mean `F.binary_cross_entropy_with_logits`  instead of `F.cross_entropy` in the last line?

**Reply by rusty1s** [reply]
Yeah, you need `binary_cross_entropy`. The `softmax` needs to be applied beforehand.
